### PR TITLE
[CALCITE-2712] Add rule to remove unnecessary outer join

### PR DESCRIPTION
**Pull request by chunweilei**
JIRA: https://issues.apache.org/jira/browse/CALCITE-2712

### CODE REVIEW

**Comment by chunweilei** [comment]
Updated the PR including:

- Rename the rule to XXXJoinRemoveRule
- Make all rules into top-level classes 
- Move some utility methods to RelOptUtil
- Add ProjectJoinRemoveRule
- Remove  the JIRA case URL in some test methods

File: core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinJoinRemoveRule.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        bottomJoin.getRowType().getFieldCount())) {
+      return;
+    }
+
+    // Check whether the top join uses unexpected columns.
+    final List<Integer> leftKeys = new ArrayList<>();
+    RelOptUtil.splitJoinCondition(topJoin.getLeft(), topJoin.getRight(),
+        topJoin.getCondition(), leftKeys, new ArrayList<>(),
+        new ArrayList<>());
+    if (leftKeys.stream().filter(s -> s >= leftBottomChildSize).count() > 0) {
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinJoinRemoveRule.java]: I hope the if condition part can be splitted into 2 statements.

File: core/src/main/java/org/apache/calcite/plan/RelOptUtil.java [hunk 2]
```diff
@@ -735,6 +737,95 @@ public static RelNode createCastRel(
...
+    };
+    return condition.accept(shuttle);
+  }
+
+  /**
+   * Whether it can remove join according to the aggregate.
+   *
+   * @param aggregate  Aggregate
+   * @param lowerLimit lower limit (included)
+   * @param upperLimit upper limit (excluded)
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/plan/RelOptUtil.java]: The explanation of lower/upper limit is unclear.

File: core/src/main/java/org/apache/calcite/plan/RelOptUtil.java [hunk 3]
```diff
@@ -735,6 +737,95 @@ public static RelNode createCastRel(
...
+   * @return Whether it can remove join or not
+   */
+  public static boolean removeJoin(Aggregate aggregate, int lowerLimit,
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/plan/RelOptUtil.java]: Better call it canRemoveJoin?

File: core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinJoinRemoveRule.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    final RexBuilder rexBuilder = topJoin.getCluster().getRexBuilder();
+    final RelBuilder relBuilder = call.builder();
+
+    relBuilder.push(bottomJoin.getLeft()).push(topJoin.getRight())
+        .join(topJoin.getJoinType(),
+            RelOptUtil.transformCondition(topJoin.getCondition(), rexBuilder,
+                leftBottomChildSize, offset))
+        .aggregate(relBuilder.groupKey(groupKey.build()),
+            RelOptUtil.transformAggregateCall(aggregate.getAggCallList(),
+                leftBottomChildSize, offset));
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinJoinRemoveRule.java]: How about extract some variables for debugability and readability?

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinJoinRemoveRule.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+    int offset = bottomJoin.getRight().getRowType().getFieldCount();
+    final RexBuilder rexBuilder = topJoin.getCluster().getRexBuilder();
+
+    RexShuttle shuttle = new RexShuttle() {
+      @Override public RexNode visitInputRef(RexInputRef inputRef) {
+        return rexBuilder.makeInputRef(inputRef.getType(),
+            inputRef.getIndex() - offset);
+      }
+    };
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinJoinRemoveRule.java]: I think we have some utility functions that can do the shuttle work. 
RexUtil.shift?

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinJoinRemoveRule.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+    List<RexNode> newExprs = project.getProjects().stream()
+        .map(expr -> expr.accept(shuttle)).collect(Collectors.toList());
+    final RelBuilder relBuilder = call.builder();
+
+    relBuilder.push(bottomJoin.getLeft()).push(topJoin.getRight())
+        .join(topJoin.getJoinType(),
+            RelOptUtil.transformCondition(topJoin.getCondition(), rexBuilder,
+                leftBottomChildSize, offset)).project(newExprs);
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinJoinRemoveRule.java]: Can we split this?

File: core/src/main/java/org/apache/calcite/plan/RelOptUtil.java (hunk 3 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/plan/RelOptUtil.java]: Yup.

File: core/src/main/java/org/apache/calcite/plan/RelOptUtil.java (hunk 2 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/plan/RelOptUtil.java]: Got it. Will explain it clearer.

File: core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinJoinRemoveRule.java (hunk 1 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinJoinRemoveRule.java]: The `if` condition can be simplified to `leftKeys.stream().anyMatch(s -> s >= leftBottomChildSize)`.

File: core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinJoinRemoveRule.java (hunk 4 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinJoinRemoveRule.java]: Sure.

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinJoinRemoveRule.java (hunk 6 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinJoinRemoveRule.java]: Sure.

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinJoinRemoveRule.java (hunk 5 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinJoinRemoveRule.java]: Good point. Will check it.

**Comment by chunweilei** [comment]
Update the PR including:
* Rename `removeJoin` to `canRemoveJoin`
* Use `RexUtil#shift` to transform expressions
* Rename 'transformAggregateCall' to 'shift'
* Simplify some `if` conditions to `leftKeys.stream().anyMatch(s -> s >= leftBottomChildSize)`
* Add some comments

**Comment by chunweilei** [comment]
Update the PR including:
* Remove utility methods shift and canRemoveJoin after using more general methods to indicate whether the join can be removed
* Use AggregateCall.transform to generate new aggregate calls.
* Add getAllFields to get all fields from an aggregate in RelOptUtil.java.

**Comment by chunweilei** [comment]
@julianhyde any other comments?

**Comment by julianhyde** [comment]
Too busy to review right now.

**Comment by chunweilei** [comment]
@julianhyde I am wondering if you have time to review this PR now. Appreciate it.

**Comment by chunweilei** [comment]
Squash commits.

**Pull request merged**